Mathlib/LinearAlgebra/Matrix/HermitianFunctionalCalculus.lean

```diff
@@ -0,0 +1,302 @@
+
+This file defines an instance of the continuous functional calculus for Hermitian matrices over an
+RCLike field 𝕜.
+
+## Main Results
+
+- definition of φ : the real StarAlgHom from C(spectrum ℝ A, ℝ) to (Matrix n n 𝕜) appearing in the
+                    instance.
+- instContinuousFunctionalCalculus : Instance of the Continuous functional Calculus for a hermitian
+                                     matrix A over 𝕜.
+
+## Tags
+
+spectral theorem, diagonalization theorem, continuous functional calculus
+-/
+
+section ConjugateUnits
+
+variable {R A : Type*} [CommSemiring R] [Ring A] [Algebra R A]
+
+/-- Conjugation by a unit preserves the spectrum, inverse on right. -/
+@[simp]
+lemma spectrum.conjugate_units {a : A} {u : Aˣ} :
+    spectrum R (u * a * u⁻¹) = spectrum R a := by
+  suffices ∀ (b : A) (v : Aˣ), spectrum R (v * b * v⁻¹) ⊆ spectrum R b by
+    refine le_antisymm (this a u) ?_
+    apply le_of_eq_of_le ?_ <| this (u * a * u⁻¹) u⁻¹
+    simp [mul_assoc]
+  intro a u μ hμ
+  rw [spectrum.mem_iff] at hμ ⊢
+  contrapose! hμ
+  simpa [mul_sub, sub_mul, Algebra.right_comm] using u.isUnit.mul hμ |>.mul u⁻¹.isUnit
+
+/-- Conjugation by a unit preserves the spectrum, inverse on left. -/
+@[simp]
+lemma spectrum.conjugate_units' {a : A} {u : Aˣ} :
+    spectrum R (u⁻¹ * a * u) = spectrum R a := by
+  simpa using spectrum.conjugate_units (u := u⁻¹)
+
+end ConjugateUnits
+
+section UnitaryConjugate
+
+universe u
+
+variable {R A : Type*} [CommSemiring R] [Ring A] [Algebra R A] [StarMul A]
+
+/-- Unitary conjugation preserves the spectrum, star on left. -/
+@[simp]
+lemma spectrum.unitary_conjugate {a : A} {u : unitary A} :
+    spectrum R (u * a * (star u : A)) = spectrum R a :=
+  spectrum.conjugate_units (u := unitary.toUnits u)
+
+/-- Unitary conjugation preserves the spectrum, star on right. -/
+@[simp]
+lemma spectrum.unitary_conjugate' {a : A} {u : unitary A} :
+    spectrum R ((star u : A) * a * u) = spectrum R a := by
+  simpa using spectrum.unitary_conjugate (u := star u)
+
+end UnitaryConjugate
+
+section FiniteSpectrum
+
+universe u v w
+
+/-- An endomorphism of a finite-dimensional vector space has a finite spectrum. -/
+theorem Module.End.finite_spectrum {K : Type v} {V : Type w} [Field K] [AddCommGroup V]
+    [Module K V] [FiniteDimensional K V] (f : Module.End K V) :
+    Set.Finite (spectrum K f) := by
+  convert f.finite_hasEigenvalue
+  ext f x
+  exact Module.End.hasEigenvalue_iff_mem_spectrum.symm
+
+variable {n R : Type*} [Field R] [Fintype n] [DecidableEq n]
+
+/-- An n x n matrix over a ring has a finite spectrum. -/
+theorem Matrix.finite_spectrum (A : Matrix n n R) : Set.Finite (spectrum R A) := by
+  rw [← AlgEquiv.spectrum_eq (Matrix.toLinAlgEquiv <| Pi.basisFun R n) A]
+  exact Module.End.finite_spectrum _
+
+instance Matrix.instFiniteSpectrum (A : Matrix n n R) : Finite (spectrum R A) :=
+  Set.finite_coe_iff.mpr (Matrix.finite_spectrum A)
+
+end FiniteSpectrum
+
+section SpectrumDiagonal
```

Comment:
This section should go in another file, although I'm not sure which is best off the top of my head. This should also constitute another PR.